### PR TITLE
find_single_number (Resolves #86)

### LINKED ISSUE #86: Single Number
## Description:
Given a non-empty array of integers `nums`, every element appears twice except for one. Find and return that single element.

## Input:
A list of integers `nums`, where each integer appears exactly twice, except for one.

## Output:
An integer representing the element that appears only once in the array.

## Example 1:

- **Input:** `nums = [2, 2, 1]`  
- **Output:** `1`

## Example 2:

- **Input:** `nums = [4, 1, 2, 1, 2]`  
- **Output:** `4`

## Example 3:

- **Input:** `nums = [1]`  
- **Output:** `1`

Sources

[Single Number - LeetCode](https://leetcode.com/problems/single-number/description/)

### PR DESCRIPTION
**Pull request by mudassra-taskeen**
---
Name: Review solution of find_single_number
About: A template PR for reviewing the code of find_single_number with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_find_single_number.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Tests for the `find_single_number` module.
+
+This tests the correctness of the function `find_single_number`:
+    - Different numbers of elements in the input list.
+    - Defensive assertions for invalid inputs.
+    - Boundary cases like empty and large lists.
+
+Author: Mudassra Taskeen
+Date: 2025-01-07
+"""
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py:6]: Hey Mudassara, 
This module docstring is very good but it can be improved and organized well if it contains separate categories for different test cases like this.
Test Categories:
    - Standard Cases: Example tests
    - Edge Cases: Example tests
    - Defensive Cases: Example tests
  
In this way a person can easily take an overview of what kind of test cases does this file contains in a proper way.

File: solutions/tests/test_find_single_number.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import unittest
+import sys
+import os
+
+# Add the solutions directory to the PYTHONPATH
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../solutions")))
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py]: I noticed that this this import statement can cause certain troubles like it can cause **ImportError** during execution time in different environments. Also, it's sometimes harder to debug with **sys.path** 
So you can use this path "_from solutions.find_single_number import find_single_number_" to avoid these path related issues path-related issues.
and you must have faced them while running your tests as well.

Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: You can replace line 20 and 22 with this import statement

File: solutions/tests/test_find_single_number.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        with self.assertRaises(AssertionError):
+            find_single_number([1, "2", 3])
+
+    def test_boundary_large_list(self):
+        """
+        Test boundary case for a large list.
+        """
+        nums = [i for i in range(1, 1000)] * 2 + [1001]
+        self.assertEqual(find_single_number(nums), 1001)
+
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py:50]: All the test cases are good but there should be test that handles empty list as an input. It's very important

Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py:50]: I like the test cases but I just noticed that the function takes an input list that contains only integers. So, there should be a test that handles a list with floating-point numbers as well and raises an error.
Like I just tested it by adding this print(find_single_number([1.5, 2, 2, 1.5, 3])) print statement in your function file and it raised error.

File: solutions/tests/test_find_single_number.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py:51]: I can see that all the test cases are dealing with lists that contains positive integers so their should be one or two basic test cases that deals with list containing positive and negative integers and 0 as well.
It will verify that the function behaves correctly with different ranges of inpurs.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py:50]: I've reviewed all the tests. They are all running and are correct.
There's just a minor adjustment which can be done like arranging tests following this organization.
Basic Cases: Start with the simplest, most common use cases
Edge Cases: Test boundary conditions and special values
Defensive Cases: Test input validation and error handling
Performance/Boundary Cases: Test resource-intensive scenarios 

Following this sequence, your test_boundary_large_list which is an edge case should come before test_invalid_input and test_defensive_non_integer_elements which are defensive cases.

File: solutions/tests/test_find_single_number.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    - Defensive assertions for invalid inputs.
+    - Boundary cases like empty and large lists.
+
+Author: Mudassra Taskeen
+Date: 2025-01-07
+"""
+
+import unittest
+import sys
+import os
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py]: Using this direct import method "from solutions.find_single_number import find_single_number" which I suggested below, there is no need to import sys and os separately so line 16 and 17 can be removed.

File: solutions/tests/test_find_single_number.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+# Add the solutions directory to the PYTHONPATH
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../solutions")))
+
+from find_single_number import find_single_number
+
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py]: Mudassara you did great job. Everything is good, there's just need to add consistent spacing in this test file. Like over here from line 23 to 26 there are all extra empty space lines. So they can be replaced with one single empty space line.

File: solutions/tests/test_find_single_number.py (hunk 1 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:6]: Oh yes Fatima, you are right. I’ll update the docstring to include sections for Standard Cases, Edge Cases, and Defensive Cases, along with example tests for each category.

File: solutions/tests/test_find_single_number.py (hunk 2 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py]: I have added your recommended path but it has created an import problem

File: solutions/find_single_number.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    assert isinstance(nums, list), "Input must be a list"
+    assert all(isinstance(i, int) for i in nums), (
+        "All elements in the list must be integers"
+    )
+
+    result = 0
+    for num in nums:
+        result ^= num  # XOR operation, all duplicate numbers cancel out
+
+    return result
```
Comment (fatima-malik99) [inline solutions/find_single_number.py]: Hey Mudassara,
Your solution is good but there is a little problem the the naming of the variables according to these 3 checks in the checklist.
Variable names help to understand the strategy
Variables are named with snake_case
Variable names are clear and helpful
 
 So, I think that the current naming can be changed for example:
1. Change parameter 'nums' to 'numbers' for more clarity.
2. Rename 'result' to 'single_number' to better reflect its purpose and we can rename 'i' with 'num' then.
3. Also, you can use 'current_number' instead of 'num' in the loop.

File: solutions/find_single_number.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        >>> find_single_number([2, 2, 1])
+        1
+
+        >>> find_single_number([4, 1, 2, 1, 2])
+        4
+
+        >>> find_single_number([1])
+        1
+
+    Example:
```
Comment (fatima-malik99) [inline solutions/find_single_number.py]: There are already 3 doctests which are very good and they are running as well so this example can be removed. Good job with the doctests.

Comment (fatima-malik99) [reply solutions/find_single_number.py]: Also it contains a print statement which is not allowed according to the given checks.

File: solutions/tests/test_find_single_number.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:6]: That's great. Let me know if any help is required.

File: solutions/tests/test_find_single_number.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: Can you share the error here?
Also remove import sys and os.

File: solutions/find_single_number.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    Examples:
+        >>> find_single_number([2, 2, 1])
+        1
+
+        >>> find_single_number([4, 1, 2, 1, 2])
+        4
+
+        >>> find_single_number([1])
+        1
+
```
Comment (fatima-malik99) [inline solutions/find_single_number.py:36]: I noticed that the function includes defensive assertions to check the validity of nums, which is great 👏 However, it would be helpful to document the possible exceptions in a Raises section of the docstring as per the checklist.

File: solutions/tests/test_find_single_number.py (hunk 5 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py]: When I ran the unit tests on the VS code, my two checks (formatting and unit tests) failed. So I did every possible way to pass those checks. That's why these imports were added. But your point is right here. Now I removed these imports, and the code is working fine.

File: solutions/tests/test_find_single_number.py (hunk 2 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py]: I have figured it out. Basically, "solutions" was causing an issue. After removing that, everything is going well.

File: solutions/find_single_number.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    Example:
+        nums = [4, 1, 2, 1, 2]
+        print(find_single_number(nums))
+        # Output: 4
+
+    Defensive Assertions:
+        - Ensures the input is a list of integers.
+    """
+    assert isinstance(nums, list), "Input must be a list"
```
Comment (fatima-malik99) [inline solutions/find_single_number.py]: These assertions are good but according to the check "Each assertion checks for only one assumption about the argument," there's room for improvement. Right now, some assertions combine multiple checks into one, making debugging harder if an error occurs. So here assertions can be added like this:

```
assert isinstance(nums, list), "Input must be a list"
assert len(nums) > 0, "List cannot be empty"
assert all(isinstance(i, int) for i in nums), "Each element must be an integer"
```

File: solutions/find_single_number.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    assert all(isinstance(i, int) for i in nums), (
+        "All elements in the list must be integers"
+    )
+
+    result = 0
+    for num in nums:
+        result ^= num  # XOR operation, all duplicate numbers cancel out
+
```
Comment (fatima-malik99) [inline solutions/find_single_number.py:58]: Actually, I liked the implementation, it's well-structured and concise but I guess that the code could be spaced a bit better to improve readability and clearly separate different stages of execution.

**Review by fatima-malik99 — CHANGES_REQUESTED** [review]
Hi Mudassara, 
You actually did a very good job. I've just added a few suggestions, you can have a look and change your files accordingly.

File: solutions/tests/test_find_single_number.py (hunk 4 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:71]: Alright. Now I have added three extra tests:
- Positive+negative integers
- Zero+positive+negative integers
- All negative integers 

It will verify the behavior of the function when it comes to dealing with different types of inputs.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:50]: Oh, okay, I have a lot of work to do. I am considering your valuable suggestions and working on it.

Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:50]: I didn't notice that. I'm grateful for highlighting this error. Now I have added a test, which contains a float number, and it's also running smoothly. You can also check on your side.

File: solutions/tests/test_find_single_number.py (hunk 2 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:20]: I have figured it out. Basically, "solutions" was causing an issue. After removing that, everything is going well.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:70]: I didn't notice that. I'm grateful for highlighting this error. Now I have added a test, which contains a float number, and it's also running smoothly. You can also check on your side.

File: solutions/tests/test_find_single_number.py (hunk 4 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:71]: Oh, okay, I have a lot of work to do. I am considering your valuable suggestions and working on it.

Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:51]: Alright. Now I have added three extra tests:

Positive+negative integers
Zero+positive+negative integers
All negative integers
It will verify the behavior of the function when it comes to dealing with different types of inputs.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:50]: I have created a test for this but this is raising an error

File: solutions/tests/test_find_single_number.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: That's great then Mudassara, Keep going.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: Can you share that test with me?

Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: Alright, let me check.

File: solutions/tests/test_find_single_number.py (hunk 4 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:51]: Good job Mudassara, let me have a look.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: That's fine Mudassara, take it easy. Let me know if you need any help.

File: solutions/tests/test_find_single_number.py (hunk 5 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: I am glad to hear that. I am so happy that you have resolved this issue.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: Yes, I've checked them they all are good and running. Great job

Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: Now all the categories are well organized Mudassara. You did a great job.

File: solutions/tests/test_find_single_number.py (hunk 6 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py]: I wanted to differentiate the import statement from the rest of the code, that's why I did this. But you are right here. It looks more professional if the spacing is equal throughout the code.

File: solutions/find_single_number.py (hunk 7 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: I completely forgot about the important checklist points here. I preferred short and concise names over full names. But yeah, considering the three important checks, I have changed the short names with the full names so that meaning and purpose should be clear.

File: solutions/find_single_number.py (hunk 8 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: I didn't understand this point, Fatima. Could you please elaborate it? I didn't add a print statement.

File: solutions/find_single_number.py (hunk 9 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py:36]: Fatima, Please give an example so that I can look into it.

File: solutions/find_single_number.py (hunk 10 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: I put these assertions in my function file, but it created an error. When I added some tests for empty lists, they failed due to a type error. After that, I have to add different modifications to both files. You can review it and let me know if you want any changes further.

File: solutions/find_single_number.py (hunk 11 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py:58]: I didn't notice that. Now all code lines are equally spaced.

File: solutions/tests/test_find_single_number.py (hunk 6 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: Yes, Mudassara. I've just checked now your code is spaced evenly. Well done!

File: solutions/find_single_number.py (hunk 7 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py]: That's completely fine. You know when we solve leetcode questions so we are mostly habitual of using short keyword variables. Now I've checked all the variables are good.

File: solutions/find_single_number.py (hunk 8 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py]: No you actually didn't added any print statement inside the code but this example is unnecessary and it contains a print statement as well so it can be removed.

File: solutions/find_single_number.py (hunk 9 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py:36]: Sure, you can add like this.
    Raises:  
        AssertionError: If input is not a list
        AssertionError: If any element is not an integer

File: solutions/find_single_number.py (hunk 10 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py]: That was not an error. These assertions are supposed to raise error when input is unexpected so don't worry about it. You can add these anyways.

File: solutions/find_single_number.py (hunk 11 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py:58]: Yeah, I saw that. Everything is good now.

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py:50]: Sure. I have added this test in my file as an edge case on line number 29
    # Edge Cases
    def test_empty_list(self):
        with self.assertRaises(AssertionError):
            find_single_number([])

File: solutions/find_single_number.py (hunk 8 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: I have a total 3 doctests, so this one is not extra. This is the third one. You can check that.

File: solutions/find_single_number.py (hunk 9 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py:36]: Okay, Fatima, I have added that to the file.

File: solutions/find_single_number.py (hunk 10 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: Alright. I got your point. I have added these three assertions in my function file.

File: solutions/find_single_number.py (hunk 8 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py]: Hey mudassara, I am talking about line no# 38, 39 and 40. That's not a doctest actually

**Comment by mudassra-taskeen** [comment]
I have added all of the recommended changes Fatima, I am truly grateful for your time and valuable suggestions. Highly appreciated.

File: solutions/find_single_number.py (hunk 9 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py:36]: Yes, I've checked that. It's great.

File: solutions/tests/test_find_single_number.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        with self.assertRaises(AssertionError):
+            find_single_number([1, "2", 3])
+
+    # Performance/Boundary Cases
+    def test_boundary_large_list(self):
+        nums = [i for i in range(1, 1000)] * 2 + [1001]
+        self.assertEqual(find_single_number(nums), 1001)
+
+
+if __name__ == "_main_":
```
Comment (fatima-malik99) [inline solutions/tests/test_find_single_number.py]: Hi mudassara, 
there should be 2 underscores with main
`if __name__ == "__main__":`

File: solutions/tests/test_find_single_number.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py:50]: There is no error with this test case. I've just checked all the test are running and working fine. Good job!

File: solutions/find_single_number.py (hunk 10 above)
Comment (fatima-malik99) [reply solutions/find_single_number.py]: Great! I have checked them.

File: solutions/tests/test_find_single_number.py (hunk 12 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_find_single_number.py]: yeah I have rectified it.

Comment (fatima-malik99) [reply solutions/tests/test_find_single_number.py]: That's great. It's good now.

File: solutions/find_single_number.py (hunk 8 above)
Comment (mudassra-taskeen) [reply solutions/find_single_number.py]: Oh yeah, I have not noticed that before . I removed that example where there was a print statement.

Comment (fatima-malik99) [reply solutions/find_single_number.py]: Now it's perfect. Good job.

**Review by fatima-malik99 — APPROVED** [review]
Hi Mudassara, you did a great job overall. I've reviewed your code and all the suggestions have been added. Solution is ready to merge now.
